[BE]예약 추가하기 기능 구현

choitree opened the pull request:
(#11) post_booking 기능 구현에 대한 pr입니다.
중간 중간에 pull 한 코드랑, 삭제되지 않은 .idea 파일 등으로 지저분할 수 있는데 감안하고 봐주세요😭

user에서 booking이 list로 있는 구조에서, booking에서 user를 가지는 구조로 변경되어서 작성이 수월했네요
코드가 너무 단순해서 이게 맞나 싶긴한데 확인 부탁드려용🥰

호텔은 1개 샘플로 넣었는데, 실제 정보로 집어넣었습니다!! 이거 언제 다 넣져...?ㅎㅎㅎㅎㅎㅎ

bong6981 committed:
chore: merge를 위한 커밋(#8)

bong6981 committed:
chore: 머지를 위한 커밋(#8)

bong6981 committed:
feat: Contorller 분리(#16)

- 컨트롤러를 분리한다

bong6981 committed:
feat: gitignore 파일에 build 폴더 추가(#16)

- build폴더를 gitignore 파일에 포함

bong6981 committed:
feat: gitignore 파일 수정

- 트리가 맥을 사용하여gitignore파일 수정

choitree committed:
chore : Room클래스에 변수명 수정

choitree committed:
chore : 추적되고 있는 불필요한 .class 파일 삭제

bong6981 committed:
feat: entity내 파일들을 폴더로 묶음(#16)

bong6981 committed:
feat: entity 관계 설정(#16)

- User가 booking과 whishlist를 List로 가질 수 있도록 관계 설정

bong6981 committed:
feat: 관계적용을 한 스키마 변경(#16)

- 테이블간의 관계를 list로 묶어준 내용을 반영하여 스키마 sql 변경

bong6981 committed:
feat: entity 변경(#16)

- 합의한 내용에 따라 entity변경

bong6981 committed:
Merge branch 'be/feat/16/get_booking' of https://github.com/dahun-lee-daji/airbnb into be/feat/16/get_booking

bong6981 committed:
feat: 스키마 sql수정(#16)

- 필드에 빠진 부분이 있어 스키마 sql수정

bong6981 committed:
feat: entity와 sql 일치하지 않은 부분 수정(#16)

bong6981 committed:
feat: Booking 관계변경(#16)

- Booking이 User의 List로 들어가있던 구조에서
- userId를 필드로 갖는 것으로 변경

bong6981 committed:
feat: Booking DAO에서 findById, save 구현(#16)

choitree committed:
chore : .idea 디렉토리 내 파일캐시 삭제 불가능으로 커밋

choitree committed:
chore : 추적되고 있는 불필요한 파일 삭제

choitree committed:
Feat : post_booking 기능 구현(#11)

choitree committed:
Feat : post_booking 기능 구현(#11)

choitree committed:
chore : 오타 수정 (#11)

choitree committed:
chore : 16브랜치에서 pull한 코드 커밋

choitree committed:
chore : 16브랜치에서 pull한 코드 커밋

choitree committed:
chore : 16브랜치에서 pull한 코드 커밋

choitree committed:
chore : BookingRequestDTO 수정

- 기본 생성자 추가
- allargs 생성자 삭제
- @JsonProperty 추가
  -totalPrice, roomId에 붙이지 않으면 오류 발생

choitree committed:
chore : BookingDAO 수정

- 주석 삭제

bong6981 commented on be/airbnb/src/main/java/com/team19/airbnb/AirbnbApplication.java:
```diff
@@ -9,5 +9,4 @@
 	public static void main(String[] args) {
 		SpringApplication.run(AirbnbApplication.class, args);
 	}
-
```
여기 개행 삭제해 주시면 좋을 것 같아요! 저도 파일 공백 정리를 안하고 pr을 보내서.. 저희 다음에 pr 보낼 때 주의하고자 개행 부분에 다 코멘트 남기겠습니다. 😄

choitree replied:
저는 파일 마지막줄에 개행이 필요하다고 알고있어서 설정을 개행 추가하는 것으로 해뒀는데
노트북 바꾸고서는 안해놓은거같긴하네요ㅠㅠ

중간 중간 개행은 저희 레거시대로 작성하고, 마지막 1줄은 남겨둬야하는 것으로 알고있습니당~~ 혹시 제가 잘못 알고있다면 말씀해주세용😊
https://github.com/codesquad-members-2021/java-chess/pull/9#discussion_r572027344

bong6981 commented on be/airbnb/src/main/java/com/team19/airbnb/domain/Booking.java:
```diff
@@ -0,0 +1,97 @@
+
+    private Long user;
+    private Long room;
+
```
요기에 개행이 있네요! 😢  아마 제 실수일 것 같아요!

choitree replied:
개행 수정했습니당

bong6981 commented on be/airbnb/src/main/java/com/team19/airbnb/domain/Booking.java:
```diff
@@ -0,0 +1,97 @@
+                user, room);
+    }
+
+    public static Booking create(LocalDate checkIn, LocalDate checkOut,
+                                 Integer guest, BigDecimal totalPrice,
+                                 Long room) {
+        return new Booking(null,
+                checkIn, checkOut,
+                guest, totalPrice,
+                null, room);
+    }
```
이건 어떨 때 필요한 것인가용 *_* ?

choitree replied:
requestDTO에서 toEntity메소드를 만들게되면, 해당 dto를 엔티티로 변환해주어야하는데
그 때 해당 dto에 존재하는 것과 동일한 구성의 생성자가 필요합니당
근데 우리는 생성자 대신 create 메소드를 사용하기로 해서 저렇게 만들었습니다!!
저기서 userId를 null로 설정하고, checkId를 Booking에서 따로 확인하고 추가하는 방식으로 로직이 구성됩니당

setId이지만, checkId를 사용한 것은 현재는 setter랑 동일한 기능의 로직이지만,
향후 오오쓰를 반영하면 변경될 수 있어서 우선 메소드명을 checkId로 설정해두었습니당!!

bong6981 replied:
오오 제가 미처 파악하지 못했습니다. 저 부분에 대한 생성자도 만들어야 겠네요 !*_* 다음 브랜치에서 만들어보겠습니다!

bong6981 commented on be/airbnb/src/main/java/com/team19/airbnb/domain/Booking.java:
```diff
@@ -0,0 +1,97 @@
+    public static Booking create(LocalDate checkIn, LocalDate checkOut,
+                                 Integer guest, BigDecimal totalPrice,
+                                 Long user, Long room) {
+        return new Booking(null,
```
이 부분 제가 만든 것인데! 원래 생성자를 하나만 쓰고 싶어서 null로 전달했는데 DAO를 쓰면서 기본생성자가 필요해졌습니다. 그래서 자동으로 생성자가 여러개라 JDBC에서 인식할  생성자에 @PersistenceConstructor 를 붙였습니다. 그래서 여러 개의 생성자를 만들어도 이제 무방할 것 같아요! 제가 이부분 저 16브랜치에서 고치겠습니다!

bong6981 commented on be/airbnb/src/main/java/com/team19/airbnb/domain/Booking.java:
```diff
@@ -0,0 +1,97 @@
+        return room;
+    }
+
+    public void checkId(Long id) {
```
setId를 바꾸신 것 같은데 왜 이름을 checkId로 하셨는지 궁금합니다! 저는 bookingDAO에서 findById의 결과로 booking을 반환할 때 해당 booking에 id를 만들어주기 위해 setId를 썼는데 checkId 는 id를 체크한다는 의미같아서요 어떤 의미로 사용하셨는지 궁금합니다❓

choitree replied:
앗 봉프가 setId를 사용하시는거였군여!! 제가 모르고 이름 자체를 변경해버렸네여😱
setId이지만, checkId를 사용한 것은 현재는 setter랑 동일한 기능의 로직이지만,
향후 오오쓰를 반영하면 변경될 수 있어서 우선 메소드명을 checkId로 설정해두었습니당!!

bong6981 commented on be/airbnb/src/main/java/com/team19/airbnb/domain/User.java:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    public static User create(String github) {
+        return new User(null, github,
```
여기도 제가 작성한 것인데 Booking과 통일하기 위해 null 전달을 없애고 생성자를 새로 만들겠습니다. (자꾸 제 스스로를 리뷰하게 되네요 🚀 )

choitree replied:
봉프 코드와 제 코드가 혼재된 상황이라서 같이 pr해주시는거 좋은거같아요!!👍👍👍👍

bong6981 commented on be/airbnb/src/main/java/com/team19/airbnb/domain/Wishlist.java:
```diff
@@ -0,0 +1,20 @@
+    }
+
+    public static Wishlist create(Long room) {
+        return new Wishlist(null, room);
```
여기도... 🤖

bong6981 commented on be/airbnb/src/main/java/com/team19/airbnb/dto/BookingRequestDTO.java:
```diff
@@ -1,31 +1,30 @@
-        this.totalPrice = totalPrice;
+    public BookingRequestDTO() {}
+
+    public Booking toEntity() {
```
👍

bong6981 commented on be/airbnb/src/main/java/com/team19/airbnb/repository/BookingDAO.java:
```diff
@@ -0,0 +1,75 @@
+public class BookingDAO {
+
+    private final JdbcTemplate jdbcTemplate;
+    private final NamedParameterJdbcTemplate namedParameterJdbcTemplate;
```
namedParameterJdbcTemplate 은 현재 사용되고 있지 않는 것 같습니다!

choitree replied:
네네네 현재는 사용되지 않는데, 지우도록 할게용~~

bong6981 commented on be/airbnb/src/main/java/com/team19/airbnb/repository/BookingDAO.java:
```diff
@@ -0,0 +1,75 @@
+
+    public void bookReservation(Booking booking, Long userId) {
+        String query = "INSERT INTO booking (check_in, check_out, guest, total_price, user, room) VALUES (?, ?, ?, ?, ?, ?)";
+        jdbcTemplate.update(query, booking.getCheckIn(), booking.getCheckOut(), booking.getGuest(), booking.getTotalPrice(), userId, booking.getRoom());
```
booking 자체에 userId 필드가 있기 때문에 서비스 단에서 먼저 booking에 userId를 넣어준 상태에서 resposiotry로 들어와야 하지 않나 싶어요! (서비스의 역할이라고 생각합니다!) 

위의 save 메소드와 역할이 똑같아서 booking으로 받으면 위의 save 메소드를 사용하면 될 것 같습니다!

choitree replied:
봉프 이거 고치다가, 제가 엄청난 잘못을 햇다는걸 알게되었네요!!
봉프가 만들어놓은 setId는 booking id이고, 
제가 사용하려고 한 메소드는 setUserId 였는데 잘못 수정한거였어요ㅠㅠㅠㅠ
죄송합니다 이거 원래 의도대로 돌려놓고 setId도 다시 만들어놓을게요!

bong6981 commented on be/airbnb/src/main/java/com/team19/airbnb/service/BookingService.java:
```diff
@@ -0,0 +1,36 @@
+        Booking booking = bookingDAO.findById(bookingId).orElseThrow(IllegalStateException::new);
+    }
+
+    public void createBooking(@RequestBody BookingRequestDTO bookingRequestDTO, Long userId) {
```
여기에 @requestBody를 붙이는 이유가 궁금합니다!

choitree replied:
아까 널포인터 오류떳을때 이거저거 해봤던 잔해였는데 지워도 작동이 잘되네요😦
지우도록 하겠습니다 꼼꼼하게 리뷰해주셔서 감사해여~~ 제가 좀 덜렁덜렁거려서 잘 부탁드립니당🥰🥰
다음 pr에는 좀 더 체크하고 보내볼게용!!!!

bong6981 replied:
저도 덜렁 덜렁.. *_* 저도 잘 부탁드립니다!

bong6981 commented on be/airbnb/src/main/resources/data.sql:
```diff
@@ -0,0 +1,4 @@
+INSERT INTO user(id, github) VALUES(1, 'tree');
+
+INSERT INTO room(id, name, grade, reviewer, address, latitude, longitude, price_per_day, room_type, room_configuration, description, host_name, host_image)
```
👏 ..... 저희 언제 다 넣죠 ㅠㅠ

choitree replied:
😭😭😭😭😭😭몇개 넣을지 목표를 정해볼까여?ㅋㅋㅋㅋㅋㅋ

bong6981 reviewed (changes requested):
크게는 
1. BookingDAO에서 save 메소드가 있는데 이를 활용하시지 않고 bookReservation을 만드신 이유가 궁금합니다 ( userId를 booking에 넣어주는 것은 서비스 단에서 이루어 지는 것이라고 생각합니다!) 

2. sedId를 왜 checkId라고 하셨는지 궁금합니다! id를 세팅한다는 의미 같은데 검사한다는 의미랑 다른 것 같아서요 (제가 영어를 잘 모르는 걸까요.. 😢 ) 

고생 많으셨습니다. 🌴

choitree committed:
chore : PR 리뷰 반영

-Booking 개행 삭제
-checkId 메소드 setId로 변경
-checkUserId 메소드 생성

choitree committed:
chore : PR 리뷰 반영

-BookingService에 createBooking메소드 수정
 - parameter에 @RequestBody삭제
 - checkId를 checkUserId로 변경

choitree committed:
chore : PR 리뷰 반영

-BookingDAO에 사용하지 않는 namedParameter 삭제
-봉프의 save 메소드에 잘못 수정한 코드 되돌림
 -checkId를 setId로 변경

choitree commented:
봉프 스피디한 PR 감사합니당 

코멘트 작성해주신 내용에 대해서 말씀드리자면, save메소드가 없는 상태에서 작성중인 메소드가 있어서 
해당 메소드로 bookReservation으로 변경했습니다.
userId를 booking에 넣는 로직을 작성했다고 생각했는데, 그게 checkUserId여야 하는걸, 봉프가 만드신 setId를 수정해서 꼬였는데
그 부분을 수정하고 DAO에서는 booking만 파라메터로 받는 방식으로 수정하였습니다.

그리고 setId는 위에 제 의견을 말씀드렸는데 혹시 잘못되었거나 봉프는 다른 생각이시면 말씀해주세용~~~
일단 null 포함된 봉프 16브랜치에 반영하시겠다고 말씀주신 부분 말고는 리뷰 반영했습니다!!
